modules/openy_features/openy_prgf/modules/openy_prgf_class_sessions/openy_prgf_class_sessions.module

```diff
@@ -0,0 +1,217 @@
+<?php
+
+/**
+ * @file OpenY Paragraph Class Sessions module file.
+ */
+
+use Drupal\openy_session_instance\Entity\SessionInstance;
+use Drupal\openy_session_instance\SessionInstanceManager;
+use Drupal\node\Entity\Node;
+use Drupal\Core\Link;
+use Drupal\Core\Url;
+
+/**
+ * Implements hook_theme().
+ */
+function openy_prgf_class_theme() {
+  $items = [
+    'class_sessions' => [
+      'variables' => [
+        'session_instances' => [],
+      ],
+      'template' => 'openy-class-sessions',
+    ],
+  ];
+  return $items;
+}
+
+/**
+ * Implements hook_preprocess_paragraph__bundle().
+ */
+function openy_prgf_class_sessions_preprocess_paragraph__class_sessions(&$variables) {
+  // Cache context.
+  $variables['#cache']['contexts'][] = 'url.query_args';
+
+  $node = Drupal::routeMatch()->getParameter('node');
+  if (!$node || $node->bundle() !== 'class') {
+    return;
+  }
+
+  // Current node is a class.
+  $class_id = $node->id();
+
+  // Get query param location.
+  $request = Drupal::request();
+
+  // Current date as timestamp.
+  $current_date = strtotime("today UTC");
+
+  /* @var $session_instance_manager Drupal\openy_session_instance\SessionInstanceManager */
+  $session_instance_manager = Drupal::service('session_instance.manager');
+
+  // Use the advanced drupal_static() pattern, since this is called very often.
+  static $drupal_static;
```

Comment:
I believe this code should be already in `openy_session_instance` module. There's couple of managers. If there's no method to do this, this code should be moved to `openy_session_instance` module and be more generic.

This function is too long and should be splitted out to different methods.